### ISSUE TITLE
CONV2D, DENSE layer with default QAT can't generate correct int8 quantized nodes in TFLite model

### ISSUE THREAD

**Issue by psunn** [tensorflow/model-optimization#523]
**Describe the bug**
when converting a MNIST model using QAT, quantize selective `CON2D` or `DENSE` layer with `tfmot.quantization.keras.quantize_annotate_layer` can't generate correct int8 quantized TFLite model.

check [colab](https://colab.research.google.com/drive/1kyUt2kRgjUmwGZ_N0825C6vO0jRPCRBF?usp=sharing) for bug reproduce

QAT only CON2D layer:
in generated tflite model (quantized_conv2d_mnist.tflite), nodes following `CON2D` layer been quantized
![image](https://user-images.githubusercontent.com/55463253/90668999-d490bf00-e248-11ea-8e30-8da13f06ad16.png)

QAT only Dense layer:
in generated tflite model (quantized_dense_mnist.tflite), nodes following `DENSE` layer been quantized
![image](https://user-images.githubusercontent.com/55463253/90668554-28e76f00-e248-11ea-8b13-81531646ec05.png)

> tfmot.quantization.keras.quantize_annotate_layer

**System information**

TensorFlow version (installed from source or binary): 2.4.0-dev20200819

TensorFlow Model Optimization version (installed from source or binary): 0.4.1

**Comment by psunn** [comment]
@alanchiao can you have a look please?

**Comment by alanchiao** [comment]
@nutsiepully 

I'd like to given that this is an important flow but have a backlog of issues I need to address first.

**Comment by Tessil** [comment]
I looked a bit into the issue and it seems the problem comes from TFMOT and not TFLite (there is another similar issue in the TF repository https://github.com/tensorflow/tensorflow/issues/45887).

From what I understand, when a layer is annotated with `quantize_annotate_layer` a `FakeQuantWithMinMaxVars` node is added after the annotated layer during the `quantize_apply` call. A `FakeQuantWithMinMaxVars` is also [added](https://github.com/tensorflow/model-optimization/blob/3a0c22d778c6d7c06e81ab1771957deba7a3d52a/tensorflow_model_optimization/python/core/quantization/keras/default_8bit/default_8bit_transforms.py#L531) implicitly after the `InputLayer`.

When a model is converted with TFLite, the converter checks for each layer if there is a fake quantization node for both the input and output as it needs the range of both to quantize a layer. If either one is missing this fake quant node, the layer is not quantized.

This works fine when we quantize the whole model or if we quantize the `N` first nodes without any gap (after the activation and batch normalization have been merged). A problem arises though when we want to quantize some nodes in the middle of the network.

For example the following
`InputLayer -> quantize_annotate_layer(Conv2D) -> quantize_annotate_layer(Conv2D) -> quantize_annotate_layer(Conv2D) -> Dense`
will be transformed to something like
`InputLayer -> FakeQuantWithMinMaxVars  -> Conv2D -> FakeQuantWithMinMaxVars -> Conv2D -> FakeQuantWithMinMaxVars  -> Conv2D -> FakeQuantWithMinMaxVars -> Dense`
and all the `Conv2D` will be quantized correctly.

On the other hand if we don't quantize the second convolution like this
`InputLayer -> quantize_annotate_layer(Conv2D) -> Conv2D -> quantize_annotate_layer(Conv2D) -> Dense`
we will have
`InputLayer -> FakeQuantWithMinMaxVars  -> Conv2D -> FakeQuantWithMinMaxVars -> Conv2D -> Conv2D -> FakeQuantWithMinMaxVars -> Dense`

When the TFLite converter encounters the third `Conv2D` layer, it will see that even though there is a fake quantization node for the output of the layer, there isn't any for the input. It need the range of the input to add a `Quantize` node to quantize the floating-point output of the second `Conv2D` to integer input. The third `Conv2D` will thus not be quantized even though it was annotated for quantization.

The problem in the original post is due to the `Reshape` layer.
`InputLayer -> Reshape -> quantize_annotate_layer(Conv2D) -> Dense`
will be converted to something like
`InputLayer -> Reshape -> Conv2D -> FakeQuantWithMinMaxVars  -> Dense`

As the input of the `Conv2D` layer has no fake quantization node, the layer will not be quantized by the TFLite converter.

Removing the `Reshape` and using `InputLayer(input_shape=(28, 28, 1))` will work. Adding `if isinstance(layer, tf.keras.layers.Reshape): return tfmot.quantization.keras.quantize_annotate_layer(layer)` to `apply_quantization_to_conv2d` would also work.

To resolve the problem we could during the `quantize_apply` call add `FakeQuantWithMinMaxVars` in front of an annotated layer too (as long as one is not already present, if the previous layer was also annotated for example). Though further discussion and investigation may be needed.

Model example:
```python
model = tf.keras.Sequential(
    [
        tf.keras.layers.InputLayer(input_shape=(28, 28, 1), batch_size=1),
        # Next Conv2D will not be quantized if we add this layer
        # tf.keras.layers.Lambda(lambda x: tf.identity(x)),
        tfmot.quantization.keras.quantize_annotate_layer(
            tf.keras.layers.Conv2D(
                filters=12, kernel_size=(3, 3), activation=tf.nn.relu
            )
        ),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        # Last Conv2D will not be quantized if we don't annotate this layer too
        tfmot.quantization.keras.quantize_annotate_layer( 
            tf.keras.layers.Conv2D(
                filters=12, kernel_size=(3, 3), activation=tf.nn.relu
            )
        ),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tfmot.quantization.keras.quantize_annotate_layer(
            tf.keras.layers.Conv2D(
                filters=12, kernel_size=(3, 3), activation=tf.nn.relu
            )
        ),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(10, activation=tf.nn.softmax),
    ]
)
```

Note that the example in https://www.tensorflow.org/model_optimization/api_docs/python/tfmot/quantization/keras/quantize_apply also doesn't seem to work for the same reason:
```python
model = keras.Sequential([
    layers.Dense(10, activation='relu', input_shape=(100,)),
    quantize_annotate_layer(layers.Dense(2, activation='sigmoid'))
])

# Only the second Dense layer is quantized.
quantized_model = quantize_apply(model)
```

**Comment by daverim** [comment]
Hi,

Sorry for the delay. I am submitting some code to make this less confusing, but for now a workaround is
```
from tensorflow_model_optimization.quantization.keras import quantizers

class DenseOutputOnlyQuantizeConfig(tfmot.quantization.keras.QuantizeConfig):

  def __init__(self):
    self.activation_attrs = ['activation']
    self.activation_quantizer = quantizers.MovingAverageQuantizer(
        num_bits=8, per_axis=False, symmetric=False, narrow_range=False)
    
  def get_weights_and_quantizers(self, layer):
    return []

  def set_quantize_weights(self, layer, quantize_weights):
    pass

  def get_activations_and_quantizers(self, layer):
   return [(getattr(layer, activation_attr), self.activation_quantizer)
            for activation_attr in self.activation_attrs]

  def set_quantize_activations(self, layer, quantize_activations):
    for activation_attr, activation in zip(
        self.activation_attrs, quantize_activations):
      setattr(layer, activation_attr, activation)
  
  def get_output_quantizers(self, layer):
    return [self.activation_quantizer]

  def get_config(self):
    return {}

  @classmethod
  def from_config(cls, config):
    return cls(**config)

model = keras.Sequential([
    quantize_annotate_layer(
        keras.layers.Dense(10, activation='relu', input_shape=(100,)), 
        DenseOutputOnlyQuantizeConfig()),
    quantize_annotate_layer(layers.Dense(2, activation='relu'))
])

# Only the second Dense layer is quantized.
with tfmot.quantization.keras.quantize_scope(
    {'DenseOutputOnlyQuantizeConfig': DenseOutputOnlyQuantizeConfig}):
  quantized_model = quantize_apply(model)
```
